log/src/main/java/net/md_5/bungee/log/ColouredWriter.java

```diff
@@ -1,56 +1,151 @@
     }
 
-    private static ReplacementSpecification compile(ChatColor color, String ansi)
+    private static Character getChar(ChatColor color)
     {
-        return new ReplacementSpecification( Pattern.compile( "(?i)" + color.toString() ), ansi );
+        return color.toString().charAt( 1 );
+    }
+
+    private static final String formatAnsi(String msg)
+    {
+        StringBuilder stringBuilder = new StringBuilder();
+        char[] chars = msg.toCharArray();
+        for ( int index = 0; index < chars.length; index++ )
+        {
+            char currentChar = chars[ index ];
+            if ( currentChar == ChatColor.COLOR_CHAR )
+            {
+                if ( index == chars.length - 1 )
+                {
+                    return stringBuilder.append( ChatColor.COLOR_CHAR ).toString();
+                }
+                char nextChar = chars[ ++index ];
+
+                if ( Character.toLowerCase( nextChar ) == 'x' && index + 12 < chars.length )
+                {
+                    StringBuilder hex = new StringBuilder( "#" );
+                    for ( int j = 0; j < 6; j++ )
+                    {
+                        int cur = index + 2 + ( j * 2 );
+                        if ( chars[ cur - 1 ] != ChatColor.COLOR_CHAR )
+                        {
+                            break;
+                        }
+                        hex.append( chars[ cur ] );
+                    }
+                    try
+                    {
+                        // Ansi can't support all colors, so we get a close one
+                        ChatColor closest = getClosestChatColor( ChatColor.of( hex.toString() ).getColor().getRGB() );
+                        //  Hex color codes are resetting special effects like boldness
+                        stringBuilder.append( ANSI_COLOR_MAP.get( 'r' ) ).append( ANSI_COLOR_MAP.get( getChar( closest ) ) );
+                        index += 12;
+                        continue;
+                    } catch ( IllegalArgumentException ex )
+                    {
+                    }
+                }
+
+                // default color support
+                String ansi = ANSI_COLOR_MAP.get( nextChar );
+                if ( ansi != null )
+                {
+                    stringBuilder.append( ansi );
+                } else
+                {
+                    stringBuilder.append( ChatColor.COLOR_CHAR ).append( nextChar );
+                }
+            } else
+            {
+                stringBuilder.append( currentChar );
+            }
+        }
+
+        return stringBuilder.toString();
     }
 
-    private static final ReplacementSpecification[] REPLACEMENTS = new ReplacementSpecification[]
+    // from ViaBackwards https://github.com/ViaVersion/ViaBackwards/blob/master/common/src/main/java/com/viaversion/viabackwards/protocol/protocol1_15_2to1_16/chat/TranslatableRewriter1_16.java
```

Comment:
GPL code!!! Absolutely not allowed in Bungee.

Reply:
Damn